src/vm/contexts.rs

```diff
@@ -224,11 +515,25 @@ impl <'a> GlobalContext <'a> {
     
     pub fn new(database: ContractDatabase<'a>) -> GlobalContext<'a> {
         GlobalContext {
+            parent_map: None,
             database: database,
-            read_only: false
+            read_only: false,
+            asset_map: AssetMap::new()
         }
     }
 
+    pub fn log_asset_transfer(&mut self, sender: &PrincipalData, contract_name: &str, asset_name: &str, transfered: Value) {
+        let asset_identifier = AssetIdentifier { contract_name: contract_name.to_string(),
+                                                 asset_name: asset_name.to_string() };
+        self.asset_map.add_asset_transfer(sender, asset_identifier, transfered)
+    }
+
+    pub fn log_token_transfer(&mut self, sender: &PrincipalData, contract_name: &str, asset_name: &str, transfered: i128) -> Result<()> {
```

Comment:
I'm having a hard time understanding why these methods exist: `log_asset_transfer` and `log_token_transfer`.  Isn't it the case that an asset or token transfer should be immediately materialized in the DB as part of the current checkpoint, so that subsequent asset or token transfers on the same token or asset type operate on the latest state of the asset or tokens affected?  Why track any state in the environment outside the DB?

Reply:
The AssetMap is used to track the transfers that occur in a single transaction. The database doesn't actually track that information -- it only a materialized key-value store of:

Assets: `AssetKey -> Owner`
Tokens: `Principal -> Balance`

So even though these operations are atomic, reversable, they don't encode the actual transfer. That's only implicit in the database.

The AssetMap is what will allow us to evaluate whether or not the whitelist has been violated by a transaction.

Reply:
What I'm trying to figure out is what happens under the following circumstances.  Suppose Alice has 10 stackaroos, and suppose there exists a contract C that can own stackaroos -- it can receive and send stackaroos on behalf of a caller.  Can Alice send 10 stackaroos to C, and then have C send 10 stackaroos to Bob all in a single transaction (i.e. with `as-contract`)?  If so, then how does the VM know that C temporarily has 10 stackaroos in-between when Alice sends them to C and when Alice calls the method to send them from C to Bob?

Reply:
Ultimately, there would be two invocations of `token-transfer!` --

The first would log: 

`stackaroos, Alice, 10`

The second would log 

`stackaroos, C, 10`

To make sure that these logs bubble up from child contexts, the `.commit()` function in the context combines its entries to its parent context.

Reply:
Gotcha -- it was unclear to me where the `.commit()` was being called.  Is there a test for this kind of scenario somewhere I can play around with?

Reply:
There's integration tests for this in `vm/tests/assets.rs` -- the name transfer around line 395 tests to make sure entries are added for the name transfer and the token transfer (the transfer names function intentionally does a token transfer as well to test this functionality) -- there's a couple of other transfers in that file where there are also some asserts to make sure that they show up in the final assetmap (and that aborted transfers _do not_ show up).

There's some unit tests in `vm/contexts.rs` which exercise the combination of AssetMaps directly.

Reply:
Ah, I think I understand the source of my confusion.  The `AssetMap` doesn't record pending balances or asset ownerships.  It merely records that asset/token transfers _happened_.  The `AssetMap` is calculated on a per-transaction basis, and is done in the service of verifying that the user-given transaction post-conditions hold.

I'm working through the rules for this in SIP 005 right now.  Would it be fair to say that a transaction's post-conditions ought to be limited to only the principals that could conceivably transfer tokens during the transaction's execution? In my example above, that would mean Alice's transaction could put constraints on whether or not `Alice` or `C` transferred tokens or assets.  Or, would you prefer that whitelists are constrained to just the transaction sender's account (i.e. `Alice`, but not `C`)?  The reason I ask is because I don't know what the computational complexity would be of validating the set of accounts a transaction has post-conditions for -- for example, would it be allowed for a transaction have a post-condition on `Bob` even though `Bob` doesn't show up in this table at all?

Also, what do you think these post-conditions would look like?  I was envisioning a conjunction of boolean statements about the absolute quantity of tokens/assets for a given account (e.g. `Alice.stackaroos >= 1000 && C.xyzzy < 10`).  What did you have in mind?

Reply:
If we allow queries on absolute balances (or on whether or not an account _owns_ an asset) at the end of a transaction, would the `AssetMap` be good enough to tell us this?

Reply:
I imagined something like simply a list of:

```
Entry := (principal, token name, token amount) | (principal, asset name, asset identifier)
```

Where this list was treated as a whitelist for any specified principal. I'm hesitant to think that we should support any more complicated predicates, for fear of creating complex predicates that are more difficult to support in wallets, are more difficult for end users to understand (ultimately the predicates are for users not developers).

As far as predicates on the actual balance goes, no, the current AssetMap couldn't support that.

Reply:
I agree that predicates should be limited to just the spending account for a transaction, if only because it's guaranteed to be loaded into RAM already as part of checking its tx fee balance (and thus very cheap to examine).  Agreed that the predicate should not involve other accounts in order to keep the complexity down.

I'm not sure if we should rule out predicates on the _state_ of the spending account, however.  I think we should consider supporting checks on both _spending events_ and _account state_.  For example, if I'm executing a smart contract that allows me to buy a BNS name for some STX, then the predicate should be something like "the spending account now owns this name" instead of "a name was sent to the spending account".  The reason is that the _event_ "a name was sent to the spending account" does _not_ reflect the post-condition the user cares about, since it is possible that the user calls a malicious smart contract method to buy the name -- the transaction code the user executes could transfer the name from the seller to their account, and then transfers both the STX and the name from their account back to the seller.  A predicate about their _account state_ would prevent such a transaction from executing, but a predicate about their _spending event_ would allow it.

Reply:
> I'm not sure if we should rule out predicates on the state of the spending account, however. I think we should consider supporting checks on both spending events and account state. For example, if I'm executing a smart contract that allows me to buy a BNS name for some STX, then the predicate should be something like "the spending account now owns this name" instead of "a name was sent to the spending account". The reason is that the event "a name was sent to the spending account" does not reflect the post-condition the user cares about, since it is possible that the user calls a malicious smart contract method to buy the name -- the transaction code the user executes could transfer the name from the seller to their account, and then transfer both the STX and the name from their account back to the seller. A predicate about their account state would prevent such a transaction from executing, but a predicate about their spending event would allow it.

This sounds much more like a use case for allowing user transactions to contain clarity code to me, rather than including predicates in the wire format-- whether or not you own a name after executing a function is pretty straight-forward to test in clarity. So, if the plan is to support transactions with raw clarity code, then we probably shouldn't support any predicates apart from on transfers.

Reply:
I'm also not convinced we wouldn't want predicates on principals other than the sender -- a user that is transacting through a smart contract wallet may wish to only spend so much out of that contract principal.

Reply:
> This sounds much more like a use case for allowing user transactions to contain clarity code to me, rather than including predicates in the wire format-- whether or not you own a name after executing a function is pretty straight-forward to test in clarity. So, if the plan is to support transactions with raw clarity code, then we probably shouldn't support any predicates apart from on transfers.

This appears to contradict my model of your understanding of how transactions work.  Right now, I think that you think that there are only two transaction types -- one for defining a smart contract (which has the side-effect of also instantiating a smart contract account), and one for calling a method in an existing smart contract (and does not instantiate a new smart contract account).  Are you saying that we need to expand the second transaction type to be able to execute any Clarity code?

I'm strongly of the opinion that our transaction wire format should offer first-class support for post-conditions on the transaction's originating account _outside of_ the Clarity code.  In my SIP 005 draft, the rationale is:

```
Logically speaking, the post-conditions in a transaction are _representable_ as a single
Clarity statement that is evaluated immediately after the successful evaluation of the
transaction code body.  The post-conditions would be representable as a single
conjunction of a sequence of boolean statements containing only asset queries,
comparators, and literals.  This conjunction would be nested within an `(if
...)` statement that would evaluate an `(err ...)` if the conjunction was false
(i.e. at least one post-condition was false), thereby causing the transaction to
abort.

However, because we consider post-conditions to be such a vital safety feature,
post-conditions are _encoded_ in the transaction _outside of_ the Clarity code
body.  In particular, they are encoded in a compressed format that losslessly
decompresses to the above Clarity codes snippit.  Thus, post-conditions do not
pose a violation of the Stacks design principle of showing all code "as-is" in the
blockchain wire format.

The reasons for encoding post-conditions this way are three-fold.  First, by
treating post-conditions as a first-class transaction component, the protocol guarantees
that they will be applied regardless of the contained Clarity code.  While the
Clarity code can have its own statements that implement _de-facto_
post-conditions, the act of placing post-conditions on the originating account's
assets outside the code body guarantees that they will always be run.
Second, by treating them as a separate field, we make it much easier for wallet 
developers to encode them -- namely, the wallet does not need to consider 
or even access the Clarity code body to process post-conditions.  Third, 
encoding post-conditions separately from the code body allows us to make
them cheaper to store and evaluate than had they been defined as Clarity code.
```

> I'm also not convinced we wouldn't want predicates on principals other than the sender -- a user that is transacting through a smart contract wallet may wish to only spend so much out of that contract principal.

100% agreed that there will be cases where we will want this.  However, I think that post-conditions on a transaction's originating account (which is usually also the spending account) will be _so common_ and offer such a huge security win that it's worth it to make them as cheap as possible to evaluate and as simple as possible to encode.  In particular:

* It should be as simple as possible for a wallet to transform the post-conditions into a GUI of check-boxes and form fields and back
* It should be possible for a block explorer to represent the post conditions as a terminating snippit of Clarity code in the transaction
* It should be as cheap as possible to evaluate post-conditions (they can't be free) in order to encourage wide-spread use.

Reply:
> Right now, I think that you think that there are only two transaction types -- one for defining a smart contract (which has the side-effect of also instantiating a smart contract account), and one for calling a method in an existing smart contract (and does not instantiate a new smart contract account). Are you saying that we need to expand the second transaction type to be able to execute any Clarity code?

If this is an important use case, then yes, I believe we should.

> I'm strongly of the opinion that our transaction wire format should offer first-class support for post-conditions on the transaction's originating account outside of the Clarity code. In my SIP 005 draft, the rationale is:

I see. I think the disagreement here stems from viewing this as a post-condition. In my mind, the goal is to mimic for first-class assets the guarantees that any blockchain provides for its native token: maximum transfer guarantees at the wire format level -- not a general purpose ability to provide post-conditions (which is where things can get tricky, because we'll be fighting the urge to define a new domain-specific language for encoding post-conditions, and trying to draw lines on what kinds of data may be inputs to the post-conditions). If all we're doing is providing maximum transfer guarantees, then a lot of those design choices become much simpler.

> 100% agreed that there will be cases where we will want this. However, I think that post-conditions on a transaction's originating account (which is usually also the spending account) will be so common and offer such a huge security win that it's worth it to make them as cheap as possible to evaluate and as simple as possible to encode

Right -- this is an example of why we might want to provide a just a whitelist. It's relatively easy to encode different principals in a whitelist. If we wanted general post-conditions, supporting conditions on multiple principals would start to look very much like a new programming language.

Reply:
> I see. I think the disagreement here stems from viewing this as a post-condition. In my mind, the goal is to mimic for first-class assets the guarantees that any blockchain provides for its native token: maximum transfer guarantees at the wire format level -- not a general purpose ability to provide post-conditions (which is where things can get tricky, because we'll be fighting the urge to define a new domain-specific language for encoding post-conditions, and trying to draw lines on what kinds of data may be inputs to the post-conditions). If all we're doing is providing maximum transfer guarantees, then a lot of those design choices become much simpler.

I think I have my answer to my original question -- the purpose of the `log_*()` methods and the `AssetMap` are to prevent double-spending of assets.  That we could use this information to enforce post-conditions is a serendipitous side-effect, but not the intended goal of these methods and structs.

That said, I stand by my proposal -- I think there's more than enough evidence from the plethora of lost/stolen funds from smart contracts that users would greatly benefit from explicitly applying post-conditions on their accounts' assets.  I also think that the fact that this is still a problem today despite language-level support for post-conditions means that language-level support is not good enough -- post conditions must be a first-class transaction concern, and must be applied _outside of_ the language runtime.

To answer your immediate concerns, I'm willing to commit to _not_ implementing a DSL -- I am only concerned with making boolean statements about the quantity/presence of assets in the originating account (i.e. the account that sends the transaction; not necessarily the account that pays the transaction fee).  Each post-condition is a triple -- an asset name, a comparator, and a literal value to compare it to.  I don't care to support anything else beyond this, since the whole point of this feature as I understand it is to help the principal who sends the transaction remain safe from bad contracts.  I firmly believe that complex expressions on account assets, or statements about other accounts' assets, should be done within the Clarity code-body.

Reply:
>  Each post-condition is a triple -- an asset name, a comparator, and a literal value to compare it to. I don't care to support anything else beyond this, since the whole point of this feature as I understand it is to help the principal who sends the transaction remain safe from bad contracts.

I disagree with this -- it doesn't fit the mental model of a transfer. For a real world metaphor, imagine a check: you write the amount you want to send on the check, _not_ your final balance. Other blockchains more or less all work like this -- they tell you how much you are sending, _not_ your final balance.

There's also a lot of weird behavior if your balance _changes_ due to some other transaction between when you sign and when the transaction is included in a block -- suddenly the post-condition might fail, even though the intended transaction would have been successful and safe (i.e., you have enough balance to cover the transfer, and you still want to have transferred, but the final balance is no longer equal to what you expect).

This is why I think we should be encoding a list of maximum spends, e.g., a list of triples

```
(principal, asset_type, max_amount_spent)
```

> I think I have my answer to my original question -- the purpose of the log_*() methods and the AssetMap are to prevent double-spending of assets. That we could use this information to enforce post-conditions is a serendipitous side-effect, but not the intended goal of these methods and structs

Not exactly -- the `log_*` methods exist only to encode 'transfer events' in the assetmap, which exists solely to answer questions like "How many Stacks did I transfer during the current transaction?" and "Which names did I transfer during the current transaction?" -- that can be used to enforce the "transfer limits" on transactions, but it won't prevent double-spends (that's prevented by the balance/ownership checks in `assets.rs`)

Reply:
> I disagree with this -- it doesn't fit the mental model of a transfer. For a real world metaphor, imagine a check: you write the amount you want to send on the check, not your final balance. Other blockchains more or less all work like this -- they tell you how much you are sending, not your final balance.

On the contrary, in our blockchain, each account has a linearized history of operations.  Your balance before and after each transaction is imminently knowable.

> There's also a lot of weird behavior if your balance changes due to some other transaction between when you sign and when the transaction is included in a block -- suddenly the post-condition might fail, even though the intended transaction would have been successful and safe (i.e., you have enough balance to cover the transfer, and you still want to have transferred, but the final balance is no longer equal to what you expect).

Yes, I can see how that would happen.  However, given my BNS name-sale scam contract attack above, I don't think that this invalidates the use-case for them.  I think that the right answer here could be for allowing for checks on _changes_ as well as checks on _values_.  I very much want to be sure that I own the BNS name at the end of the transaction, for example, and checking for a change doesn't let me do this.

Reply:
>>I disagree with this -- it doesn't fit the mental model of a transfer. For a real world metaphor, imagine a check: you write the amount you want to send on the check, not your final balance. Other blockchains more or less all work like this -- they tell you how much you are sending, not your final balance.

>On the contrary, in our blockchain, each account has a linearized history of operations. Your balance before and after each transaction is imminently knowable.

I'm not sure I follow. Sure, there's a linearized history of operations, but that's true of pretty much any system designed to handle asset transfers. My point still stands -- any time you are reasoning about a transfer, you almost always reason about the delta, _not_ the final balance. Real world examples of fungible asset transfers are things like wire transfers and checks, all of which encode deltas. When thinking about non-fungible items, reasoning about final balance is even more difficult. If I'm selling a car, how would I represent a post-condition by final balance? The predicate would be "my car balance does not contain 'sold car', but must contain 'complete list of other cars I should still own at the end of the transaction'" 

> I very much want to be sure that I own the BNS name at the end of the transaction, for example, and checking for a change doesn't let me do this.

But that logic could apply to _any_ kind of smart contract invocation -- not just token/asset transfers (for example, lets say I make a contract for paying to update a bitmap, I'd want to be sure that the bitmap updates if I pay). If I'm calling a function to perform an atomic name purchase, that function _should implement that correctly_. Generally speaking, its the contracts responsibility to ensure that functionality like that is implemented correctly.

Rather, the goal of whitelisting transfers is to ensure that contracts aren't surreptitiously sneaking off with tokens you didn't expect them to. Note that this is the same goal as the design of allowances in ERC20. If someone writes a bad name swap contract, and I call it, fully expecting to pay into it, then we should allow that.

Reply:
> I'm not sure I follow. Sure, there's a linearized history of operations, but that's true of pretty much any system designed to handle asset transfers. 

I was pointing out that we _could_ do post-conditions total balance if we wanted to, since I thought you were saying that we could not (but I think your point was that doing so is just awkward in practice since no one does this in the real world?)

Also, the post-condition of "I sold car X" would be encoded as "car X is _not_ in my account's 'car' asset map".  Similarly, the post-condition of "I buy car X" would be encoded as "car X is in my account's 'car' asset map".

> But that logic could apply to any kind of smart contract invocation -- not just token/asset transfers (for example, lets say I make a contract for paying to update a bitmap, I'd want to be sure that the bitmap updates if I pay). If I'm calling a function to perform an atomic name purchase, that function should implement that correctly. Generally speaking, its the contracts responsibility to ensure that functionality like that is implemented correctly.

Okay, this is where I think we disagree philosophically.  I think we both agree on the premise that in the world of smart contracts, smart contract bugs are permanent -- you can't upgrade smart contracts or deactivate malicious smart contracts, so you can't fix bugs or stop scam developers.  I think we also agree that the solution to mitigating bugs and stopping scammers in this world is to empower the _user_ to block the execution of smart contracts that trigger bugs or give money or assets to scammers.

Where we disagree is that I don't think simply whitelisting accounts or setting a maximum outbound transfer of assets goes far enough towards empowering users to survive in this world.  Users not only need to do this, but also need to make sure everything is okay once their transaction completes.

In your bitmap example, if it is possible for the bitmap pixels to be encoded as a first-class asset, then I think it's actually quite reasonable to empower the user to block the transaction if their requested pixels aren't set once they pay.

>  If I'm calling a function to perform an atomic name purchase, that function should implement that correctly.  Generally speaking, its the contracts responsibility to ensure that functionality like that is implemented correctly.

Yes, to a first approximation.  But, the user doesn't care as to whether or not they lost their funds due to a bug or due to a scam.  They care that they lost their funds.  I don't think any blockchain today goes nearly far enough to help make sure this can't happen, and I think we actually have a fighting chance to address this.

> Note that this is the same goal as the design of allowances in ERC20.  If someone writes a bad name swap contract, and I call it, fully expecting to pay into it, then we should allow that.

Why would you willingly call a bad name swap contract?  Presumably, you don't do so willingly -- you want your name in exchange for your tokens.  So, how do you make sure that you'll get your name?  In the absence of an out-of-band way to sue the developer for fraud (which is what would happen in the real world), you'll at the very least need a way to make sure the "bad outcome" does not happen.  You could say that it's up to everyone to independently audit the name swap code to tell _a priori_ that it's buggy, but that's just blaming the user.  My proposal addresses this -- you'd say that the name must be in your asset table once the transaction finishes, or the transaction aborts.  Allowances do not catch this.

Reply:
Okay, we talked a bit more.  The key takeaway is that transaction post-conditions should apply to transfers in the AssetMap, and to the account states of non-fungible tokens, but nothing more.  Anything more complex will go into the Clarity code body.